src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -554,6 +554,56 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
   CallGenerator::do_late_inline_helper();
 }
 
+// replace box node to scalar node only in case it is directly referenced by debug info
+static void replace_box_to_scalar(CallNode* call, Node* resproj) {
+  if (resproj != nullptr && call->is_CallStaticJava() &&
+      call->as_CallStaticJava()->is_boxing_method()) {
+    Unique_Node_List debuginfo_node_list;
+    for (DUIterator_Fast imax, i = resproj->fast_outs(imax); i < imax; i++) {
+      Node* m = resproj->fast_out(i);
+      if (m->is_SafePoint()) {
+        SafePointNode* sfpt = m->as_SafePoint();
+        uint dbg_start = sfpt->is_Call() ? sfpt->as_Call()->tf()->domain()->cnt() : (uint)TypeFunc::Parms+1;
+        for (uint i = 0; i < dbg_start; i++) {
+          if (sfpt->in(i) == resproj) {
+            return;
+          }
+        }
+        debuginfo_node_list.push(m);
+      } else {
+        return;
+      }
+    }
+
+    GraphKit kit(call->jvms());
+    PhaseGVN& gvn = kit.gvn();
+    // delay box in runtime, treat box as a scalarized object
+    while (debuginfo_node_list.size() > 0) {
+      ProjNode* res = resproj->as_Proj();
+      Node* debuginfo_node = debuginfo_node_list.pop();
+
+      ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
+      int n_fields = klass->nof_nonstatic_fields();
+      assert(n_fields == 1, "the klass must be an auto-boxing klass");
+
+      uint first_ind = (debuginfo_node->req() - debuginfo_node->jvms()->scloff());
+      Node* sobj = new SafePointScalarObjectNode(gvn.type(res)->isa_oopptr(),
+#ifdef ASSERT
+                                                  call->isa_Allocate(),
+#endif // ASSERT
+                                                  first_ind, n_fields, true);
+      sobj->init_req(0, kit.root());
+      debuginfo_node->add_req(call->in(res->_con));
```

Comment:
I don't understand why you are selecting the input based on the result projection field `res->_con`?

Reply:
Thank you for your review. I use `sfpt->add_req(call->in(TypeFunc::Parms));` instead of this.